### ISSUE TITLE
Figure out how to handle vector inputs to distribution constructors appropriately

### ISSUE THREAD

**Issue by alexpghayes** [alexpghayes/distributions3#38]
Suppose you want to simulate from a logistic regression model. You calculate the linear predictor, and now you need `n` (number of observation) *different* `Bernoulli` objects, and want to draw a single sample from each of them. This is one place where the base R approach shines. Should we vectorize object construction? Then we lose type stability because the type of the output (distribution vs list<distribution>) depends on the length of the input.

Feedback and possible solutions welcome.

**Comment by statibk** [comment]
I think you can already accomodate vectorization by simply passing vectors of parameters:

```
R> d <- Normal(0:3, 2)
R> quantile(d, 0.5)
[1] 0 1 2 3
```

It would probably make sense, though, to employ internally a data.frame() or a matrix() of parameters instead of a list(). Then parameters would automatically get expanded to n x p where n is the sample size and p the number of parameters. This would make subsetting easier etc.

So far so good. But you would probably expect something different from

```
R> quantile(d, c(0.025, 0.975))
[1] -3.919928  4.919928 -1.919928  6.919928
```

Presumably, what was intended here is rather something like:

```
R> sapply(c(0.025, 0.975), quantile, d = d)
          [,1]     [,2]
[1,] -3.919928 3.919928
[2,] -2.919928 4.919928
[3,] -1.919928 5.919928
[4,] -0.919928 6.919928
```

Except possibly if the length of the quantiles vector is the same as the number of distributions. Then possibly you really want also a vector output, not a matrix.

Finally, it could be handy to make the `"distribution"` object behave like a vector. Then you could put it as a column into a data.frame (or tibble), e.g., as the output of a predict() function. Thus, you would have distributional forecasts from which you can then compute quantities you are interested in. In the case above this would mean that with d[2] you could select the Normal(1, 2) distribution and length(d) should give 4 etc.

**Comment by alexpghayes** [comment]
Vectorizing distribution objects are now available in https://github.com/mitchelloharawild/distributional/, although I'm not clear how far along the package is on the whole. We could either mimic the `vctrs` approach there, which I'm sure is well done, or perhaps try to merge packages, taking the object structure from `distributional` and the documentation from `distributions3`. @mitchelloharawild any wisdom or thoughts?

**Comment by mitchelloharawild** [comment]
{distributional} is coming along quickly, it is now usable as model output with nice accessors for statistics of interest. I'm particularly excited for its integration with the {ggdist} package (https://github.com/mitchelloharawild/distributional/issues/24).

The next necessary steps for the project are:
- **Performance**: The main complexity with vectorisation comes when different distributions types are combined. Currently this is done inefficiently by storing each distribution in its own list, however the next release will vectorise within distribution classes (https://github.com/mitchelloharawild/distributional/issues/25).
- **Functionality**: The core functionality for using distributions as prediction output is complete, but some other functions more useful for working with distributions are needed. This includes functions for the moments, (log) likelihoods, skewness, etc.
- **Documentation**: The documentation in this package is much better, especially with the distribution's functions/parameterisation detailed in function's `@details`. This is something I'd like to have in the package, and merging this in would be greatly appreciated. A vignette and dynamically generated distribution characteristics table is also needed.

If you'd like to add `vctrs` support to this package for representing a vector of `Bernoulli` objects, I'd suggest restricting the distributions in the vector to be a single type. Otherwise I would suggest that {distributional} is better suited for mixing distribution classes in a vector.

**Comment by alexpghayes** [comment]
Definitely should use `distributional` for vectorized distribution objects now. We should probably add an informative error message and error on vectorized inputs, or something along those lines.

**Comment by mnlang** [comment]
I've been thinking about this post and you could indeed vectorize distribution objects of the same distribution with relatively few adaptations. Regardless of the fact that this is quite nice for teaching, this would definitely be interesting for other applications as well - especially since `distributions3` has so few dependencies. 

In my forked repo within the branch [vectorize_distr](https://github.com/mnlang/distributions3/tree/vectorize_distr), I obtained the following results with minimal changes for the normal distribution:
```
R> (d <- Normal(0:3, 2))
[1] "Normal distribution (mu = 0, sigma = 2)"
[2] "Normal distribution (mu = 1, sigma = 2)"
[3] "Normal distribution (mu = 2, sigma = 2)"
[4] "Normal distribution (mu = 3, sigma = 2)"
R> mean(d)
[1] 0 1 2 3
R> variance(d)
[1] 4 4 4 4
R> skewness(d)
[1] 0 0 0 0
R> kurtosis(d)
[1] 0 0 0 0
R> random(d)
[1] -4.208800  3.248514  1.810552  2.970799
R> random(d, 3)
            r_1      r_2        r_3
[1,] -0.5223547 2.850667 -0.5311626
[2,]  1.6746693 4.452493  0.3313276
[3,] -0.6115062 3.241150  1.9195451
[4,] -0.1449963 4.991660  5.5387037
R> pdf(d, 0)
[1] 0.1994711 0.1760327 0.1209854 0.0647588
R> log_pdf(d, 0)
[1] -1.612086 -1.737086 -2.112086 -2.737086
R> cdf(d, 0.5)
[1] 0.5987063 0.4012937 0.2266274 0.1056498
R> quantile(d, 0.5)
[1] 0 1 2 3
R> quantile(d, c(0.5, 0.5))
     q_0.5 q_0.5_1
[1,]     0       0
[2,]     1       1
[3,]     2       2
[4,]     3       3
R> quantile(d, c(0.1, 0.5, 0.9))
          q_0.1 q_0.5    q_0.9
[1,] -2.5631031     0 2.563103
[2,] -1.5631031     1 3.563103
[3,] -0.5631031     2 4.563103
[4,]  0.4368969     3 5.563103
R> support(d)
      min max
[1,] -Inf Inf
[2,] -Inf Inf
[3,] -Inf Inf
[4,] -Inf Inf
```
For a single distribution, everything works as before:
```
R> d1 <- d[1]
R> random(d1)
[1] 1.333339
R> random(d1, 3)
[1] -0.299964  1.463391  2.717911
R> pdf(d1, 0)
[1] 0.1994711
R> log_pdf(d1, 0)
[1] -1.612086
R> cdf(d1, 0.5)
[1] 0.5987063
R> quantile(d1, c(0.1, 0.5, 0.9))
[1] -2.563103  0.000000  2.563103
R> support(d1)
[1] -Inf  Inf
```
Also when simulating from multiple distributions - which was the original intention of this issue - vectorization clearly beats a simple `replicate()`.
```
R> d <- Normal(rep(0, 1000))
R> fun_random <- function(n) {
+   set.seed(123)
+   random(d, n)
+ }
R> fun_replicate <- function(n) {
+   set.seed(123)
+   replicate(n, random(d))
+ }
R> microbenchmark::microbenchmark(
+   fun_random(1000),
+   fun_replicate(1000)
+ )
Unit: microseconds
                expr        min         lq        mean     median         uq
    fun_random(1000)    104.688    112.147    158.7292    140.012    148.138
 fun_replicate(1000) 107009.849 108811.302 113574.1296 109699.675 112300.892
        max neval
   1689.419   100
 173915.670   100
```
If you like the implementation, I could send you everything including some more checks via PR next week. Of course I am also very happy to respond to any suggestions from you. Currently the other distributions, a bit of code cleaning, and a few issues with some plotting functions is still needed.

**Comment by mnlang** [comment]
I have put together a suggestion for vectorized distribution objects in PR #71. I hope the examples there illustrate my approach, but I am of course still open for alternative suggestions!

**Comment by alexpghayes** [comment]
Closed in #71!
